File: codejam/python/2009/2/2_WateringPlants.py
import math
import sys
from collections import deque, namedtuple

## Input crap
infile = sys.stdin
intokens = deque()
def getTokens(): 
    while not intokens:
        for c in infile.readline().rstrip().split() :
            intokens.append(c)    
def gs(): getTokens(); return intokens.popleft()
def gi(): return int(gs())
def gf(): return float(gs())
def gbs(): return [c for c in gs()]
def gis(n): return [gi() for i in range(n)]
def ia(m): return [0] * m
def iai(m,v): return [v] * m
def twodi(n,m,v): return [iai(m,v) for i in range(n)]
def fill2(m) : r = gis(2*m); return r[0::2],r[1::2]
def fill3(m) : r = gis(3*m); return r[0::3],r[1::3],r[2::3]
def fill4(m) : r = gis(4*m); return r[0::4],r[1::4],r[2::4],r[3::4]
MOD = 998244353
##MOD = 1000000007

Center = namedtuple("Center","i j x y")

def testr(R,N,XF,YF,RF) :
    centers = []
    for i in range(N) :
        if R < RF[i] : continue
        centers.append(Center(i,i,XF[i],YF[i]))
        x1,y1,r1 = XF[i],YF[i],RF[i]
        for j in range(i+1,N) :
            x2,y2,r2 = XF[j],YF[j],RF[j]
            d = math.sqrt((x2-x1)*(x2-x1)+(y2-y1)*(y2-y1))
            if R < 0.5*(r1+d+r2) : continue
            vx1,vy1 = (x2-x1)/d,(y2-y1)/d; vx2,vy2=-vy1,vx1
            u = (d*d-(R-r2)*(R-r2)+(R-r1)*(R-r1))*0.5/d
            v = 0.00 if (R-r1)*(R-r1)-u*u < 0 else math.sqrt((R-r1)*(R-r1)-u*u)
            xa,ya = x1+u*vx1+v*vx2,y1+u*vy1+v*vy2
            xb,yb = x1+u*vx1-v*vx2,y1+u*vy1-v*vy2
            centers.append(Center(i,j,xa,ya))
            centers.append(Center(i,j,xb,yb))
    nc = len(centers)
    for i in range(nc) :
        i1,j1,xa,ya = centers[i].i,centers[i].j,centers[i].x,centers[i].y
        for j in range(i,nc) :
            i2,j2,xb,yb = centers[j].i,centers[j].j,centers[j].x,centers[j].y
            good = True
            for k in range(N) :
                if k == i1 or k == i2 or k == j1 or k == j2 : continue
                xk,yk,rk = XF[k],YF[k],RF[k]
                if (xk-xa)*(xk-xa)+(yk-ya)*(yk-ya) <= (R-rk)*(R-rk) : continue
                if (xk-xb)*(xk-xb)+(yk-yb)*(yk-yb) <= (R-rk)*(R-rk) : continue
                good = False; break
            if good : return True
    return False

def main() :
    if len(sys.argv) > 1 : global infile; infile = open(sys.argv[1],'rt')
    ## PROGRAM STARTS HERE
    T = gi()
    for tt in range(1,T+1) :
        N = gi(); xx = [gf() for i in range(3*N)]; XF,YF,RF = xx[0::3],xx[1::3],xx[2::3]
        l,r = 0.00,810.0
        eps = 0.5e-5
        while True :
            m = 0.5*(l+r)
            if m-l < eps or m-l < l * eps : break
            if testr(m,N,XF,YF,RF) :
                r = m
            else :
                l = m
        ans = 0.5*(l+r)
        print(f"Case #{tt}: {ans}")

if __name__ == "__main__" :
    main()

File: codejam/cpp/2009/2/2_WateringPlants.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pi;
#define FOR(i,a) for (ll i = 0; i < (a); i++)
#define len(x) (ll) x.size()
const ll INF = 1LL << 62;
const ll MOD = 1000000007;
//const ll MOD = 998244353;
const double PI = 4*atan(double(1.0));
vector<vector<ll>> twodi(ll N, ll M, ll v) { vector<vector<ll>> res(N); for (auto &vv : res) vv.resize(M,v); return res; }

struct center {ll i,j; double x,y; };

int main (int argc, char **argv) {
    if (argc > 1) { freopen(argv[1],"r",stdin); }
    ios_base::sync_with_stdio(false);cin.tie(0);
    // PROGRAM STARTS HERE
    ll T; cin >> T;
    for (ll tt=1;tt<=T;tt++) {
        ll N; cin >> N; vector<double> XF(N),YF(N),RF(N); FOR(i,N) cin >> XF[i] >> YF[i] >> RF[i];
        auto testr = [&](double R) -> bool {
            vector<center> centers;
            FOR(i,N) {
                if (R < RF[i]) continue;
                auto x1 = XF[i]; auto y1 = YF[i]; auto r1 = RF[i];
                centers.push_back({i,i,x1,y1});
                for (ll j=i+1;j<N;j++) {
                    auto x2 = XF[j]; auto y2 = YF[j]; auto r2 = RF[j];
                    auto d = sqrt((x2-x1)*(x2-x1)+(y2-y1)*(y2-y1));
                    if (R < 0.5*(r1+d+r2)) { continue; }
                    auto vx1 = (x2-x1)/d;
                    auto vy1 = (y2-y1)/d;
                    auto vx2 = -vy1;
                    auto vy2 = vx1;
                    auto u = (d*d-(R-r2)*(R-r2)+(R-r1)*(R-r1))*0.5/d;
                    auto xx = (R-r1)*(R-r1)-u*u;
                    auto v = xx < 0 ? 0.00 : sqrt(xx);
                    auto xa = x1 + u*vx1 + v*vx2;
                    auto ya = y1 + u*vy1 + v*vy2;
                    auto xb = x1 + u*vx1 - v*vx2;
                    auto yb = y1 + u*vy1 - v*vy2;
                    centers.push_back({i,j,xa,ya});
                    centers.push_back({i,j,xb,yb});
                }
            }
            auto nc = len(centers);
            for (ll i=0;i<nc;i++) {
                auto i1 = centers[i].i; auto j1 = centers[i].j; auto xa = centers[i].x; auto ya = centers[i].y;
                for (ll j=i;j<nc;j++) {
                    auto i2 = centers[j].i; auto j2 = centers[j].j; auto xb = centers[j].x; auto yb = centers[j].y;
                    auto good = true;
                    FOR(k,N) {
                        if (k==i1||k==i2||k==j1||k==j2) continue;
                        auto xk = XF[k]; auto yk = YF[k]; auto rk = RF[k];
                        if ((xk-xa)*(xk-xa)+(yk-ya)*(yk-ya) <= (R-rk)*(R-rk)) continue;
						if ((xk-xb)*(xk-xb)+(yk-yb)*(yk-yb) <= (R-rk)*(R-rk)) continue;
						good = false; break;
                    }
                    if (good) return true;
                }
            }
            return false;
        };
        auto l(0.0), r(1200.0);
        FOR(i,60) { auto m = 0.5*(l+r); if (testr(m)) { r = m; } else { l = m; } }
        auto ans = 0.5*(l+r);
        printf("Case #%lld: %.17g\n",tt,ans);
    }
}

